src/python/WMCore/MicroService/MSTransferor/MSTransferor.py

```diff
@@ -499,7 +517,17 @@ def makeTransferRucio(self, wflow, dataIn, dids, dataSize, grouping, copies, nod
             self.logger.info("Creating rule for workflow %s with %d DIDs in container %s, RSEs: %s, grouping: %s",
                              wflow.getName(), len(dids), dataIn['name'], rseExpr, grouping)
             try:
-                res = self.rucio.createReplicationRule(dids, rseExpr, **ruleAttrs)
+                # make decision about current workflow, if it is new request we'll create
+                # new replication rule, otherwise we'll move replication rule
+                if wflow.dataReplacement:
+                    rids = self.getRuleIdsFromDoc(wflow.getName())
+                    self.logger.info("Going to move %s rules for workflow: %s", len(rids), wflow.getName())
+                    for rid in rids:
+                        # the self.rucio.moveReplicationRule may raise different exceptions
+                        # based on different outcome of the operation
+                        res = self.rucio.moveReplicationRule(rid, rseExpr, self.quotaAccount)
```

Comment:
You are not storing the rule IDs returned by this call. In practice, you would be persisting only the very last rule, which is not the expected behavior.

Instead, you should keep record of every single NEW rule id to be returned within `transferId`

Reply:
You call `moveReplicationRule` inside a for loop, simply assigning the result to `res`.
Given that the `transferId.update(res)` is outside the loop, in practice you are storing only the last rule id.

Reply:
I'm not sure that I follow, here is how it is done:
```
def execute(self, reqStatus):
      ...
      for reqSlice in grouper(requestRecords, 100):
           ...
           for wflow in reqResults:
                 ...
                success, transfers = self.makeTransferRequest(wflow, rseList)
....

def makeTransferRequest(self, wflow, rseList):
      ....
      for dataIn in wflow.getDataCampaignMap():
             ...
             success, transferId = self.makeTransferRucio(wflow, dataIn, dids, didsSize, grouping, copies, rses)

def makeTransferRucio(self, wflow, dataIn, dids, dataSize, grouping, copies, nodes):
      ....
      res = self.rucio.moveReplicationRule(rid, rseExpr, self.quotaAccount)
      ....
      transferId.update(res)
```

So, unless you point directly to a specific `for` loop you're talking about everything is in side the for (multiple) loops and all workflows are preserved in transferId list.

Reply:
discussed over zoom